### PR TITLE
tvOS support

### PR DESCRIPTION
**Pull request by davidkraus**
note that there is no webbrowser or a webview on tvOS so you will have limited functionality.

### CODE REVIEW

**Comment by p2** [comment]
Oh this is cool. One nitpick, can you change capitalization to `tvOS`?

**Comment by p2** [comment]
Thanks, but I cannot pull this until you've merged `develop` into your branch again.

**Comment by ddengler** [comment]
I just updated to the latest version and tried to match your updated project structure (hope I caught everything). We are still testing this. We had to switch back to the submodule as it is also not tvOS compatible (podspec issue only). I would submit a pull request to the original author, but the project seems to be dead. How do you intend to handle this in general? Let me know if we can help.

**Comment by p2** [comment]
Does this compile? Note that the methods you override have been changed, they now throw, and you don't need `authorizeEmbeddedFrom`. Implementing the first two are enough:

https://github.com/p2/OAuth2/blob/master/Sources/iOS/OAuth2%2BiOS.swift#L33

The _SwiftKeychain_ submodule is giving you trouble? I'm still using a fork of it because of earlier issues, but of course can't push to CocoaPods, so it's fine to just include the sources for tvOS.  
Your changes to Podspec also make the iOS version not use the Pod, can you revert so that only OS X (as before) and tvOS directly include the Sources? Just in case one of the CocoaPods users of the module on iOS depends on the setup.

**Comment by ddengler** [comment]
We will look into it. I had some CocoaPods issues installing tvOS with the submodule and iOS without in the same project. I will also open a separate issue (if none exists already) to discuss a future pod extraction for tvOS.

**Comment by p2** [comment]
Ah I see. I don't use CocoaPods myself so am not an expert, if there's a good reason to change how it's implemented we should do that.

**Comment by ddengler** [comment]
Lots of things are happening in this PR now. Sadly I was not able to split this up properly. **This should not be merged, yet**. I had do disable safariViewDelegate due to compilation errors below iOS 9. @p2 Could you have a look and let me know what you think? Why do you save the delegate anyway – it is never used? Also I removed a macro around the import to be able to test this in a real app of ours via cocoapods. Everything still works as expected on ios/tvos.

**Comment by p2** [comment]
The assignment to `safariViewDelegate` is needed so it doesn't get deallocated IIRC.

Please don't remove the `IMPORT_SWIFT_KEYCHAIN` switch and don't remove the keychain classes from the Xcode project. When compiling the project directly via Xcode OAuth2 should still compile and work. Same goes for `NO_KEYCHAIN_IMPORT` which is used during `swift build` from command line (which is not yet working).

What's the purpose of `OAuth2Account`?

**Comment by ddengler** [comment]
Understood. How do you intend to handle that safariViewDelegate does not compile on iOS8?

Will add this again and test it without cocoapods. Was just so I can test this internally.

I tried to follow the new (1.0) https://github.com/yankodimitrov/SwiftKeychain API, This is basically just a Wrapper-Class.

**Comment by p2** [comment]
Ah now I get it, you are on develop that is behind master. Apologies, I have updated develop to catch up to master. Current master uses `AnyObject` in that place so iOS 8 is still supported.

**Comment by ddengler** [comment]
@p2 can you have a look again? It is working properly within our (cocoapods) setup now and I should not have broken the use without cocoapods.

**Comment by p2** [comment]
Alright great, thanks for bearing with me. Can I ask for two favors, then I'll merge and adapt a few details:

- please use tabs for indentation and remove added whitespace (e.g. in `OAuth2AuthConfig`); easy to detect when you use GitHub's diff views
- rename the new keychain account struct `OAuth2Account` to `OAuth2KeychainAccount` and put it in its own `OAuth2KeychainAccount.swift` file. I'll factor some things out of OAuth2Base into that type, so would be good to have in its own file

Thanks a lot!

**Comment by ddengler** [comment]
Will do. Thanks.

**Comment by davidkraus** [comment]
done

File: Sources/iOS/OAuth2+iOS.swift [hunk 1]
```diff
@@ -124,7 +124,7 @@ extension OAuth2 {
 		
 		let delegate = OAuth2SFViewControllerDelegate(oauth: self)
 		web.delegate = delegate
-		authConfig.ui.safariViewDelegate = delegate
+//		authConfig.ui.safariViewDelegate = delegate
```
Comment (p2) [inline Sources/iOS/OAuth2+iOS.swift]: Any reason this is still commented out? (also on line 140)

File: p2.OAuth2.podspec [hunk 2]
```diff
@@ -26,12 +26,15 @@ Pod::Spec.new do |s|
...
-  
+  s.pod_target_xcconfig = { 'OTHER_SWIFT_FLAGS' => '-DIMPORT_SWIFT_KEYCHAIN' }
+
   s.source_files = "Sources/Base/*.swift"
   s.ios.source_files = "Sources/iOS/*.swift"
-  s.osx.source_files = "Sources/OSX/*.swift", "SwiftKeychain/SwiftKeychain/Keychain/*.swift"
+  s.osx.source_files = "Sources/OSX/*.swift"
+  s.tvos.source_files = "Sources/tvOS/*.swift"
+
+  s.dependency 'SwiftKeychain'
```
Comment (p2) [inline p2.OAuth2.podspec]: Depends on `~> 1.0`, right?

File: Sources/Base/OAuth2AuthConfig.swift [hunk 3]
```diff
@@ -44,7 +44,7 @@ public struct OAuth2AuthConfig {
 		/// Internally used to store default `NSWindowController` created to contain the web view controller.
 		var windowController: NSWindowController?
 		
-		#elseif os(iOS)
+		#elseif os(iOS)
```
Comment (p2) [inline Sources/Base/OAuth2AuthConfig.swift]: Trailing space

**Comment by p2** [comment]
Thanks! One safari delegate left but I'll fix it myself. (y)

**Comment by ddengler** [comment]
Thanks!

**Pull request merged**